src/main/java/org/graylog/aws/inputs/transports/KinesisTransport.java

```diff
@@ -90,21 +160,34 @@ public void launch(MessageInput input) throws MisfireException {
                 awsConfig,
                 authProvider,
                 nodeId,
-                graylogConfiguration.getHttpProxyUri() == null ? null : HttpUrl.get(graylogConfiguration.getHttpProxyUri())
+                graylogConfiguration.getHttpProxyUri() == null ? null : HttpUrl.get(graylogConfiguration.getHttpProxyUri()),
+                this,
+                objectMapper, configuration.intIsSet(CK_KINESIS_MAX_THROTTLED_WAIT_MS) ? configuration.getInt(CK_KINESIS_MAX_THROTTLED_WAIT_MS) : null,
+                configuration.getInt(CK_KINESIS_RECORD_BATCH_SIZE, DEFAULT_THROTTLED_WAIT_MS)
         );
 
         LOG.info("Starting Kinesis reader thread for input [{}/{}]", input.getName(), input.getId());
 
-        executor.submit(this.reader);
+        executor = getExecutorService();
+        kinesisTaskFuture = executor.submit(this.reader);
+    }
+
+    private ExecutorService getExecutorService() {
+        return Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                                                         .setDaemon(true)
+                                                         .setNameFormat("aws-kinesis-reader-%d")
+                                                         .setUncaughtExceptionHandler((t, e) -> LOG.error("Uncaught exception in AWS Kinesis reader.", e))
+                                                         .build());
     }
 
     private Consumer<byte[]> kinesisCallback(final MessageInput input) {
         return (data) -> input.processRawMessage(new RawMessage(data));
     }
 
     @Override
-    public void stop() {
-        if(this.reader != null) {
+    public void doStop() {
+        if (this.reader != null) {
+            this.stoppedDueToThrottling.set(false); // Prevent restart of consumer when input is shutting down.
```

Comment:
Should we move this outside of the `if` block to make sure it will always be set to `false` when stopping the transport?

Reply:
Agreed.